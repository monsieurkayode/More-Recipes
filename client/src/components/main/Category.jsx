import React, { Component } from 'react';
import { connect } from 'react-redux';
import debounce from 'lodash/debounce';
import PropTypes from 'proptypes';
import { Link } from 'react-router-dom';

import { searchPost } from '../../actions';
import cleanString from '../../../../shared/cleanString';
import Categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';

/**
 * @summary - Category class declaration
 * @class Category
 * @extends {Component}
 */
export class Category extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf Category
   */
  constructor(props) {
    super(props);
    this.state = {
      searchType: '',
      searchTerm: ''
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    $((this.search)).on('change', this.handleSearchType);
  }

  /**
   * @method componentWillUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('select').material_select();
  }

  /**
   * Handle search
   * @method handleSearch
   *
   * @param {void} void
   *
   * @returns {void}
   */
  handleSearch = () => {
    const { searchType, searchTerm } = this.state;
    if ((cleanString(searchTerm)).length % 3 === 0) {
      this.props.searchPost(searchType, searchTerm);
    }
  }

  /**
   * Handle change
   * @method handleChange
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  /**
   * Handle search criteria
   * @method handleSearchType
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleSearchType = (event) => {
    this.setState({
      searchType: event.target.value
    });
  }

  /**
   * @method renderCategory
   *
   * @param {number} index
   *
   * @returns {JSX} JSX
   */
  renderCategory = (index) => {
    const category = Categories[index];
    return (
      <li key={category}>
        <Link
          to={`/category/${category}`}
          className="collapsible-header"
        >
          {pascalCase(category)}
        </Link>
      </li>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const handleSearch = debounce(() => { this.handleSearch(); }, 300);
    return (
      <div className="col l2 offset-l1 m2 s12">
        <form id="search-form" onChange={handleSearch}>
          <input
            onChange={this.handleChange}
            value={this.state.searchTerm}
            className="white input-fa"
            name="searchTerm"
            type="text"
            placeholder="&#xf002; Enter search keyword..."
          />
          <select
            ref={(ref) => { this.search = ref; }}
            value={this.state.searchType}
          >
            <option value="" disabled>Search criteria</option>
            <option value="name">Title</option>
            <option value="category">Category</option>
            <option value="ingredients">Ingredients</option>
          </select>
        </form>
        <div className="category hide-on-small-only">
          <div>
            <div>
              <b>Featured Categories</b>
              <p className="divider" />
            </div>
            <ul className="collapsible z-depth-0">
              {Categories.slice(6).map(
                (category, index) => this.renderCategory(index)
              )}
            </ul>
          </div>
        </div>
      </div>
    );
  }
}

Category.propTypes = {
  searchPost: PropTypes.func.isRequired
};

export default connect(null, { searchPost })(Category);
