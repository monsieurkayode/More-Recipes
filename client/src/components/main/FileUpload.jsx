import React from 'react';
import PropTypes from 'proptypes';
/**
 * Hanldes file selection for upload
 *
 * @function handleChange
 *
 * @param {function} handler
 *
 * @returns {object} selected file
 */
const handleChange = handler => ({ target: { files } }) =>
  handler(files.length ? { file: files[0], name: files[0].name } : {});

const FileUpload = ({
  input: { onChange, onBlur, value: omitValue, ...inputProps },
  meta: omitMeta,
  ...props
}) => (
  <div id="upload-btn">
    <label htmlFor="image">
      <input
        type="file"
        id="image"
        onChange={handleChange(onChange)}
        onBlur={handleChange(onBlur)}
        placeholder="Upload Image"
        {...inputProps}
        {...props}
      />
      <span
        className="btn"
      >
        Upload Image
      </span>
    </label>
  </div>
);

FileUpload.propTypes = {
  input: PropTypes.shape({
    onChange: PropTypes.func.isRequired,
    onBlur: PropTypes.func.isRequired
  }).isRequired,
  meta: PropTypes.shape({}).isRequired
};

export default FileUpload;
