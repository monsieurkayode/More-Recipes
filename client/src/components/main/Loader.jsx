import React from 'react';
import PreloaderIcon, { ICON_TYPE } from 'react-preloader-icon';

/**
 * Loader
 * @function Loader
 *
 * @param {void} void
 *
 * @return {JSX} JSX
 */
const Loader = () => (
  <div className="loader">
    <PreloaderIcon
      type={ICON_TYPE.OVAL}
      size={80}
      strokeWidth={7}
      strokeColor="#006664"
      duration={800}
    />
  </div>
);

export default Loader;
