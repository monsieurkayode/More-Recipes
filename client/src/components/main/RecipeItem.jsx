import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'proptypes';
import pascalCase from '../../utils/pascalCase';

/**
 * RecipeItem
 * @function RecipeItem
 *
 * @param {void} void
 *
 * @return {JSX} JSX
 */
const RecipeItem = ({ recipe, upvote, downvote, grid }) =>
  (<div className={`col ${grid}`}>
    <div className="card hoverable views-small">
      <div className="card-image">
        <Link to={`/recipes/${recipe.id}`}>
          <img
            src={recipe.image}
            alt={recipe.recipeName}
          />
        </Link>
        <span className="card-title">{recipe.views} Views</span>
      </div>
      <div
        className="card-content small-cards-home"
      >
        <div className="card-title truncate">
          <strong>{recipe.recipeName}</strong>
        </div>
        <div
          className="card-details"
        >
          <Link
            to={`/category/${recipe.category}`}
            className="chip teal white-text"
          >{pascalCase(recipe.category)}
          </Link>
          <span>
            <div
              onClick={() => downvote(recipe.id)}
              className="unlike right"
            >
              <i
                className="material-icons tiny reaction"
              >thumb_down</i>{recipe.downvote}
            </div>
            <div
              onClick={() => upvote(recipe.id)}
              className="like right"
            >
              <i
                className="material-icons tiny reaction"
              >thumb_up</i>{recipe.upvote}
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>);

RecipeItem.propTypes = {
  recipe: PropTypes.shape({
    id: PropTypes.any,
    views: PropTypes.number,
    upvote: PropTypes.number,
    downvote: PropTypes.number,
    recipeName: PropTypes.string,
    category: PropTypes.string,
    ingredients: PropTypes.string,
    instructions: PropTypes.string,
    image: PropTypes.string
  }).isRequired,
  upvote: PropTypes.func.isRequired,
  downvote: PropTypes.func.isRequired,
  grid: PropTypes.string.isRequired
};

export default RecipeItem;
