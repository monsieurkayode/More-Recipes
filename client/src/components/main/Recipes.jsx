import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'proptypes';
import isEmpty from 'lodash/isEmpty';
import ReactPaginate from 'react-paginate';

import {
  upvoteAction,
  downvoteAction,
  fetchRecipesAction
} from '../../actions';
import { RecipeItem, Loader } from './';
import sad from '../../../assets/css/img/sad.png';

/**
 * @summary - Recipes class declaration
 * @class Recipes
 * @extends {Component}
 */
export class Recipes extends Component {
  /**
   * Component constructor
   * @param {void} void
   * @memberOf Recipes
   */
  constructor() {
    super();
    this.state = {
      isLoading: false
    };
  }

  /**
   * Handle page selection for pagination
   * @method handlePageClick
   *
   * @param {object} data
   *
   * @returns {void}
   */
  handlePageClick = ({ selected }) => {
    this.setState({ isLoading: true });
    const page = selected + 1;
    localStorage.setItem('currentPage', page);
    const currentPage = localStorage.getItem('currentPage');
    this.props.fetchRecipesAction(currentPage)
      .then(() => this.setState({ isLoading: false }));
  }

  /**
   * Check if there are recipes posts
   * @method hasRecipes
   *
   * @param {void} void
   *
   * @returns {boolean} boolean
   */
  hasRecipes = () => !isEmpty(this.props.recipes)

  /**
   * @method renderRecipes
   *
   * @param {number} index
   *
   * @returns {JSX} JSX
   */
  renderRecipes = (index) => {
    const recipe = this.props.recipes[index];
    return (
      <RecipeItem
        grid="l6 m10 offset-m1 s12"
        key={recipe.id}
        index={recipe.id}
        recipe={recipe}
        upvote={this.props.upvoteAction}
        downvote={this.props.downvoteAction}
      />
    );
  }

  /**
   * @method renderPagination
   *
   * @param {void} void
   *
   * @returns {JSX} JSX
   */
  renderPagination = () => {
    const { pagination: { page, pageCount } } = this.props;
    return (
      <div className="pagination-container">
        <ReactPaginate
          previousLabel={<i className="fa fa-chevron-left" />}
          nextLabel={<i className="fa fa-chevron-right" />}
          breakLabel={<a href="">...</a>}
          breakClassName={'break-me'}
          pageCount={pageCount}
          initialPage={page - 1}
          marginPagesDisplayed={2}
          pageRangeDisplayed={5}
          onPageChange={this.handlePageClick}
          disableInitialCallback
          containerClassName={'pagination'}
          subContainerClassName={'pages pagination'}
          activeClassName={'active'}
        />
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { pagination: { page, pageCount } } = this.props;
    const { isLoading } = this.state;
    return (
      <div className="col l5 m6 s12">
        <span><b>Recipes</b></span>
        <p className="divider" />
        <div className="row">
          {this.hasRecipes() ?
            <div id="recipes">
              {isLoading ? <Loader /> :
                Object
                  .keys(this.props.recipes)
                  .sort((a, b) => b - a)
                  .map(index => this.renderRecipes(index))}
            </div> :
            <div className="center-align not-found">
              <img
                src={sad}
                alt=""
              />
              <h6>
                No recipe matches your search! Try again with another keyword
              </h6>
            </div>
          }
        </div>
        {this.hasRecipes() && this.renderPagination()}
        {this.hasRecipes() &&
        <div
          className="center-align"
        >
          You are viewing page { page || 1 } of { pageCount || 1}
        </div>}
      </div>
    );
  }
}

const mapStateToProps = ({ recipes }) => ({
  recipes: recipes.recipes,
  pagination: recipes.pagination,
});

Recipes.defaultProps = {
  recipes: {},
  pagination: {}
};

Recipes.propTypes = {
  recipes: PropTypes.shape({}).isRequired,
  upvoteAction: PropTypes.func.isRequired,
  downvoteAction: PropTypes.func.isRequired,
  fetchRecipesAction: PropTypes.func.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    pageSize: PropTypes.number,
    pageCount: PropTypes.number,
    totalCount: PropTypes.number
  }),
};

export default connect(mapStateToProps,
  { upvoteAction, downvoteAction, fetchRecipesAction }
)(Recipes);
