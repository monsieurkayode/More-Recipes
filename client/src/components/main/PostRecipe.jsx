import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'proptypes';

import FileUpload from './FileUpload';
import { Loader } from './';
import { createPost, isFetching } from '../../actions';
import validate from '../../utils/validateRecipe';
import categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';
import resetPage from '../../utils/resetPage';

/**
 * @summary - PostRecipe class declaration
 * @class PostRecipe
 * @extends {Component}
 */
export class PostRecipe extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf PostRecipe
   */
  constructor(props) {
    super(props);
    this.state = {
      selectedCategory: 'others'
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    $(('#category')).on('change', this.handleCategory);
  }

  /**
   * @method componentWillUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('select').material_select();
  }

  /**
   * Handle submit
   *
   * @method onSubmit
   *
   * @param {void} void
   *
   * @returns {void}
   */
  onSubmit = () => {
    const category = this.state.selectedCategory;
    this.props.isFetching(true, 'PostRecipe');
    const values = this.props.values;
    this.props.createPost(category, values, (message) => {
      resetPage();
      this.props.history.push('/');
      setTimeout(() => Materialize.toast(message, 4000, 'grey darken-2'), 3000);
    });
  }

  /**
   * Handle category selection
   *
   * @method handleCategory
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleCategory = (event) => {
    this.setState({
      selectedCategory: event.target.value
    });
  }

  /**
   * @method renderInput
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderInput = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <input
            type={type}
            placeholder={placeholder}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderTextArea
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderTextArea = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <textarea
            placeholder={placeholder}
            type={type}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderCategory
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderCategory = (field) => {
    const { value } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <select
            id="category"
            value={value}
          >
            <option value="others">Select Category</option>
            {categories
              .map(
                category => (<option value={category} key={category}>
                  {pascalCase(category)}
                </option>)
              )
            }
          </select>
        </div>
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { handleSubmit, invalid, values, isLoading } = this.props;
    return (
      <div className="container">
        <div className="row">
          { isLoading ?
            <Loader /> :
            <form
              className="col l8 m8 s10 offset-m2 offset-l2 offset-s1"
              onSubmit={handleSubmit(this.onSubmit)}
            >
              <Field
                type="text"
                className=""
                placeholder="Enter recipe name"
                label="Recipe Name"
                name="recipeName"
                component={this.renderInput}
              />

              <Field
                name="category"
                value={this.state.selectedCategory}
                component={this.renderCategory}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Ingredients"
                name="ingredients"
                component={this.renderTextArea}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Intsructions"
                name="instructions"
                component={this.renderTextArea}
              />
              <div>
                <Field
                  name="image"
                  component={FileUpload}
                />
                <span className="image-name">
                  <p>
                    {!!values.image && values.image.name}
                  </p>
                </span>
              </div>
              <div className="form-btn-control">
                <Link className="white-text" to="/">
                  <button className="btn red">Cancel</button>
                </Link>
                <button
                  disabled={invalid}
                  type="submit"
                  className="btn"
                >
                  Post
                </button>
              </div>
            </form>}
        </div>
      </div>

    );
  }
}

PostRecipe.defaultProps = {
  values: {},
  errors: {}
};

PostRecipe.propTypes = {
  createPost: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  errors: PropTypes.shape({
    recipeName: PropTypes.string,
    ingredients: PropTypes.string,
    instructions: PropTypes.string
  }),
  values: PropTypes.shape({
    ingredients: PropTypes.string,
    instructions: PropTypes.string
  }),
  isFetching: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired
};

const mapStateToProps = ({ form, isLoading }) => ({
  values: form.PostRecipeForm.values,
  errors: form.PostRecipeForm.syncErrors,
  isLoading: isLoading.postRecipeIsLoading
});

export default reduxForm({
  validate,
  form: 'PostRecipeForm'
})(connect(mapStateToProps, { createPost, isFetching })(PostRecipe));
