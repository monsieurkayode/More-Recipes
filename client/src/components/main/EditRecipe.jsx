import React, { Component } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';
import PropTypes from 'proptypes';

import FileUpload from './FileUpload';
import Loader from './Loader';
import { editPost, fetchSingleRecipe, isFetching } from '../../actions';
import validate from '../../utils/validateRecipe';
import categories from '../../../../shared/categories';
import pascalCase from '../../utils/pascalCase';

/**
 * @summary - EditRecipe class declaration
 * @class EditRecipe
 * @extends {Component}
 */
export class EditRecipe extends Component {
  /**
   * Component constructor
   * @param {object} props
   * @memberOf EditRecipe
   */
  constructor(props) {
    super(props);
    this.state = {
      selectedCategory: '',
    };
  }

  /**
   * @method componentDidMount
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentDidMount() {
    const { recipeId } = this.props.match.params;
    $(('#category')).on('change', this.handleCategory);
    this.props.fetchSingleRecipe(recipeId);
  }

  /**
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.initialValues.category) {
      this.setState({
        selectedCategory: nextProps.initialValues.category
      });
    }
  }

  /**
   * @method componentWillUpdate
   *
   * @param {void} void
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('select').material_select();
  }

  /**
   * Handle submit
   *
   * @method onSubmit
   *
   * @param {object} values
   * @param {string} category
   *
   * @returns {void}
   */
  onSubmit = (values, category) => {
    const { recipeId } = this.props.match.params;
    category = this.state.selectedCategory;
    this.props.isFetching(true, 'EditRecipe');
    this.props.editPost(recipeId, category, values, (message) => {
      this.props.history.goBack();
      setTimeout(() => Materialize.toast(message, 4000, 'grey darken-2'), 3000);
    });
  }

  /**
   * Handle category selection
   *
   * @method handleCategory
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleCategory = (event) => {
    this.setState({
      selectedCategory: event.target.value
    });
  }

  /**
   * @method renderInput
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderInput = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <input
            type={type}
            placeholder={placeholder}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderTextArea
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderTextArea = (field) => {
    const { label, input, meta, type, className, placeholder } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <textarea
            placeholder={placeholder}
            type={type}
            {...input}
            className={className}
          />
          <label htmlFor className="active">{label}</label>
        </div>
        <div className="red-text right">
          {meta.touched ? meta.error : ''}
        </div>
      </div>
    );
  }

  /**
   * @method renderCategory
   *
   * @param {object} field
   *
   * @returns {JSX} JSX
   */
  renderCategory = (field) => {
    const { value } = field;
    return (
      <div className="row">
        <div className="input-field col-s12 l12 m12">
          <select
            id="category"
            value={value}
          >
            <option value={this.state.selectedCategory}>Select Category</option>
            {categories
              .map(
                category => (<option value={category} key={category}>
                  {pascalCase(category)}
                </option>)
              )
            }
          </select>
        </div>
      </div>
    );
  }

  /**
   * Renders the component
   * @method render
   *
   * @returns {JSX} JSX
   */
  render() {
    const { handleSubmit, invalid, initialValues, isLoading } = this.props;
    return (
      <div className="container">
        <div className="row">
          {isLoading ?
            <Loader /> :
            <form
              className="col l8 m8 s10 offset-s1 offset-m2 offset-l2"
              onSubmit={handleSubmit(this.onSubmit)}
            >
              <Field
                type="text"
                className=""
                placeholder="Enter recipe name"
                label="Recipe Name"
                name="recipeName"
                component={this.renderInput}
              />

              <Field
                name="category"
                value={this.state.selectedCategory}
                component={this.renderCategory}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Ingredients"
                name="ingredients"
                component={this.renderTextArea}
              />

              <Field
                type="text"
                placeholder="Markdown supported"
                className="materialize-textarea"
                label="Intsructions"
                name="instructions"
                component={this.renderTextArea}
              />
              <div>
                <Field
                  name="image"
                  component={FileUpload}
                />
                <span className="image-name">
                  <p>
                    {!!initialValues.image && initialValues.image.name }
                  </p>
                </span>
              </div>

              <div className="form-btn-control">
                <button
                  className="btn red"
                  onClick={() => this.props.history.push('/dashboard/recipes')}
                >
                Cancel
                </button>
                <button
                  disabled={invalid}
                  type="submit"
                  className="btn"
                >
                Update
                </button>
              </div>
            </form>}
        </div>
      </div>
    );
  }
}

EditRecipe.propTypes = {
  editPost: PropTypes.func.isRequired,
  fetchSingleRecipe: PropTypes.func.isRequired,
  initialValues: PropTypes.shape({
    id: PropTypes.number,
    views: PropTypes.number,
    upvote: PropTypes.number,
    downvote: PropTypes.number,
    recipeName: PropTypes.string,
    category: PropTypes.string,
    ingredients: PropTypes.string,
    instructions: PropTypes.string,
    image: PropTypes.shape({})
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
  }).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  invalid: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      recipeId: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  isFetching: PropTypes.func.isRequired
};

const mapStateToProps = ({ currentRecipe, isLoading }) => ({
  initialValues: {
    recipeName: currentRecipe.recipeName,
    category: currentRecipe.category,
    ingredients: currentRecipe.ingredients,
    instructions: currentRecipe.instructions,
    image: {
      file: 'empty',
      name: currentRecipe.image
    }
  },
  isLoading: isLoading.editRecipeIsLoading
});

export default compose(
  connect(mapStateToProps, { editPost, fetchSingleRecipe, isFetching }),
  reduxForm({
    validate,
    form: 'EditRecipeForm',
    enableReinitialize: true,
  }),
)(EditRecipe);
