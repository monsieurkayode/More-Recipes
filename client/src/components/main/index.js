/* jshint esversion: 6 */
import Category from './Category';
import Recipes from './Recipes';
import RecipeItem from './RecipeItem';
import TopRecipeItem from './TopRecipeItem';
import TopRecipes from './TopRecipes';
import SignupForm from './SignupForm';
import SigninForm from './SigninForm';
import TextField from './TextField';
import PostRecipe from './PostRecipe';
import EditRecipe from './EditRecipe';
import SideNav from './SideNav';
import WelcomeMessage from './WelcomeMessage';
import Loader from './Loader';

export {
  Category,
  Recipes,
  RecipeItem,
  TopRecipeItem,
  TopRecipes,
  PostRecipe,
  EditRecipe,
  SignupForm,
  SigninForm,
  TextField,
  SideNav,
  WelcomeMessage,
  Loader
};
