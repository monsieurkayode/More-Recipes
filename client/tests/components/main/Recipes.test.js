/* eslint-disable no-undef */
import React from 'react';

import { Recipes } from '../../../src/components/main/Recipes';
import data from '../../__mocks__/mockData';

const setup = (isLoading) => {
  const props = {
    recipes: {
      1: data.fetchRecipesResponse.recipes[0]
    },
    pagination: data.fetchRecipesResponse.pagination,
    isLoading,
    upvoteAction: jest.fn(),
    downvoteAction: jest.fn(),
    fetchRecipesAction: jest.fn().mockImplementation(() => Promise.resolve())
  };

  localStorage.setItem('currentPage', 1);

  const wrapper = shallow(<Recipes {...props} />);

  return {
    wrapper,
    props
  };
};

const { wrapper } = setup(false);

describe('UserRecipe', () => {
  it('should render correctly',
    (done) => {
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });

  it('should have a pagination component for page navigation', (done) => {
    expect(wrapper.find('.pagination-container').exists()).toBe(true);
    done();
  });

  it('should call handlePageClick if selected page changes', (done) => {
    const handlePageSpy = jest.spyOn(wrapper.instance(), 'handlePageClick');
    wrapper.instance().handlePageClick({ selected: 2 });
    expect(handlePageSpy).toHaveBeenCalledWith({ selected: 2 });
    done();
  });

  describe('hasRecipes method should', () => {
    let { props } = setup(false);
    props = { ...props, recipes: {} };
    const noRecipesWrapper = shallow(<Recipes {...props} />);
    it('return false if there are no recipes to display', (done) => {
      expect(noRecipesWrapper.instance().hasRecipes()).toBe(false);
      done();
    });

    it('return true if there are recipes to display', (done) => {
      expect(wrapper.instance().hasRecipes()).toBe(true);
      done();
    });
  });
});
