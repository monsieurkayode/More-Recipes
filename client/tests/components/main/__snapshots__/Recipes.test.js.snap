// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRecipe should render correctly 1`] = `
<div
  className="col l5 m6 s12"
>
  <span>
    <b>
      Recipes
    </b>
  </span>
  <p
    className="divider"
  />
  <div
    className="row"
  >
    <div>
      <RecipeItem
        downvote={[MockFunction]}
        grid="l6 m10 offset-m1 s12"
        index={1}
        key="1"
        recipe={
          Object {
            "category": "specials",
            "createdAt": "2018-02-05T07:00:47.980Z",
            "downvote": 0,
            "id": 1,
            "image": "spice.jpg",
            "ingredients": "**A list of ingredients for preparing Chicken soup**
                  * 1 (3 pound) whole chicken
                  * 4 carrots, halved
                  * 4 stalks celery, halved
                  * 1 large onion, halved
                  * water to cover
                  * salt and pepper to taste
                  * 1 teaspoon chicken bouillon granules (optional)
                  * 1/2 teaspoon lemonade",
            "instructions": "* Put the chicken, carrots, celery and onion in a large
                  soup pot and cover with cold water. Heat and simmer, 
                  uncovered, until the chicken meat falls off of the bones
                  (skim off foam every so often).
                  * Take everything out of the pot. Strain the broth. 
                  Pick the meat off of the bones and chop the carrots, celery 
                  and onion. Season the broth with salt, pepper and chicken 
                  bouillon to taste, if desired. Return the chicken, carrots, 
                  celery and onion to the pot, stir together, and serve.",
            "recipeName": "Chicken Soup",
            "upvote": 0,
            "views": 0,
          }
        }
        upvote={[MockFunction]}
      />
    </div>
  </div>
  <div
    className="pagination-container"
  >
    <PaginationBoxView
      activeClassName="active"
      breakClassName="break-me"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination"
      disableInitialCallback={true}
      disabledClassName="disabled"
      initialPage={0}
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel={
        <i
          className="fa fa-chevron-right"
        />
      }
      onPageChange={[Function]}
      pageCount={1}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel={
        <i
          className="fa fa-chevron-left"
        />
      }
      subContainerClassName="pages pagination"
    />
  </div>
  <div
    className="center-align"
  >
    You are viewing page 
    1
     of 
    1
  </div>
</div>
`;
